firebase-common/src/main/java/com/google/firebase/Timestamp.kt

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase
+
+import android.os.Build
+import android.os.Parcel
+import android.os.Parcelable
+import androidx.annotation.RequiresApi
+import java.time.Instant
+import java.util.Date
+
+/**
+ * A Timestamp represents a point in time independent of any time zone or calendar.
+ *
+ * Represented as seconds and fractions of seconds at nanosecond resolution in UTC Epoch time. It is
+ * encoded using the Proleptic Gregorian Calendar which extends the Gregorian calendar backwards to
+ * year one. Furthermore,It is encoded assuming all minutes are 60 seconds long, specifically leap
+ * seconds are "smeared" so that no leap second table is needed for interpretation. Range is from
+ * 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z. By restricting to that range, we ensure
+ * that we can convert to and from RFC 3339 date strings.
+ *
+ * @see <a href="https://git.page.link/timestamp-proto">Timestamp</a>The ref timestamp definition
+ */
+class Timestamp : Comparable<Timestamp>, Parcelable {
+  val seconds: Long
+  val nanoseconds: Int
+
+  /**
+   * Creates a new [Timestamp].
+   *
+   * @param seconds represents seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z. Must be
+   * from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59Z inclusive.
+   * @param nanoseconds represents non-negative fractions of a second at nanosecond resolution.
+   * Negative second values with fractions must still have non-negative nanoseconds values that
+   * count forward in time. Must be from 0 to 999,999,999 inclusive.
+   */
+  constructor(seconds: Long, nanoseconds: Int) {
+    validateRange(seconds, nanoseconds)
```

Comment:
In general, it's not ideal that constructors throw, as it's not expected. It would be clearer if the constructor is private and a factory method is used instead.

Reply:
I completely agree, although it's worth noting that this was already existing behavior. If we change this, I'll need to change all the existing timestamp references in our codebase to use factories instead. I'm fine with making that change- but just wanted to be clear about it.

Reply:
Keep in mind that moving Timestamp to common is actually a breaking change as well. This will be shipped alongside the other I/O breaking changes- to make this possible. 

But, the extent to which it's breaking is a different conversation. The argument that it may not be worth moving from constructors to factories- as it would require more extensive changes downstream, is a valid point.

wdyt @rlazo

Reply:
Yes, it would be breaking. 

@daymxn how breaking is this change? AFAIK moving it from firestore to common wouldn't break it since firestore has an `api` dependency in common, and since we are not changing its package, it should be seamless. Am I missing something?

Reply:
fwiw I'm OK leaving the API as is given that it's already public.

Reply:
It's a bit of weird situation to articulate, as it's a package providing an artifact for a different package.

Previously, while `Timestamp` was technically under the common package- it still lived in Firestore. So moving it to common will still be a binary change. Although, just rebuilding it would fix this as the compiler can easily infer the difference. The exception to this rule is if the dependency they pull on Firestore has an exclusion on common (as would be extremely practical if you're pulling your own dependency on common, or you're relying on a version of common necessary for another dependency).

So while it generally won't be that major of a breaking change, it *could* be depending on the consumer's existing dependency structure.